presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/optimization/ElasticSearchRowExpressionConverter.java

```diff
@@ -0,0 +1,255 @@
+        implements RowExpressionConverter<ElasticSearchConverterContext>
+{
+    private static final String EMPTY_STRING = "";
+    private static final String DEREFERENCE_SEPERATOR = "|";
+
+    public static final String COUNT_FUNCTION_NAME = "count";
+
+    protected final FunctionMetadataManager functionMetadataManager;
+    protected final StandardFunctionResolution standardFunctionResolution;
+
+    @Inject
+    public ElasticSearchRowExpressionConverter(FunctionMetadataManager functionMetadataManager, StandardFunctionResolution standardFunctionResolution)
+    {
+        this.functionMetadataManager = functionMetadataManager;
+        this.standardFunctionResolution = standardFunctionResolution;
+    }
+
+    @Override
+    public String visitCall(CallExpression call, ElasticSearchConverterContext context)
+    {
+        FunctionHandle functionHandle = call.getFunctionHandle();
+        if (standardFunctionResolution.isNotFunction(functionHandle)) {
+            return format("(NOT %s)", call.getArguments().get(0).accept(this, context));
+        }
+
+        FunctionMetadata functionMetadata = functionMetadataManager.getFunctionMetadata(functionHandle);
+
+        if (standardFunctionResolution.isOperator(functionHandle)) {
+            return handleOperator(call, functionMetadata, context);
+        }
+        return handleFunction(call, functionMetadata, context);
+    }
+
+    private String handleOperator(CallExpression call, FunctionMetadata functionMetadata, ElasticSearchConverterContext context)
+    {
+        FunctionHandle functionHandle = call.getFunctionHandle();
+
+        List<RowExpression> arguments = call.getArguments();
+
+        if (call.getArguments().size() == 1 && standardFunctionResolution.isNegateFunction(functionHandle)) {
+            String value = call.getArguments().get(0).accept(this, context);
+            String separator = value.startsWith("-") ? " " : "";
+            return format("-%s%s", separator, value);
+        }
+        if (functionMetadata.getOperatorType().get() == OperatorType.NOT_EQUAL) {
+            String variable = call.getArguments().get(0).accept(this, context);
+            String value = call.getArguments().get(0).accept(this, context);
+            return format("NOT (%s: %s)", variable, value);
+        }
+        Optional<OperatorType> operatorTypeOptional = functionMetadata.getOperatorType();
+        if (operatorTypeOptional.isPresent() && arguments.size() == 2 && (standardFunctionResolution.isComparisonFunction(functionHandle) || standardFunctionResolution.isArithmeticFunction(functionHandle))) {
+            return format(
+                    "(%s %s %s)",
+                    arguments.get(0).accept(this, context),
+                    operatorTypeOptional.get().getOperator(),
+                    arguments.get(1).accept(this, context));
+        }
+        throw new PrestoException(NOT_SUPPORTED, String.format("Unknown operator %s in push down", operatorTypeOptional));
```

Comment:
Instead of throwing an exception, we could simply call `return handleUnsupportedOptimize(context);`.
If you want to let outside know the reason of failure, you could as well log the error message or push it into context.

Same for many of the throws below.